### ISSUE TITLE
missing some generic interfaces for sorted containers

### ISSUE THREAD

**Issue by StephenVavasis** [JuliaCollections/DataStructures.jl#767]
The sorted containers do not currently support the following interface methods described in Chap. 15 of the Julia manual: reverse iteration, firstindex, lastindex.

**Comment by StephenVavasis** [comment]
Can anyone tell me: should `reverse` for the sorted containers extend `Base.reverse` or `Base.Iterators.reverse`? Or does it not matter?

**Comment by oxinabox** [comment]
I think: if it is eager and allocates a data structure that holds a complete copy of the data reversed `Base.reverse`.
But if it returns an object that can be iterated to get a reversed version of the data but is not a complete copy then `Base.Iterators.reverse`.
(e.g. if it just reverses keys but shares look up of original array of values or something)

**Comment by bjarthur** [comment]
related?

```
julia> d=Dict("a"=>1, "b"=>2)
Dict{String, Int64} with 2 entries:
  "b" => 2
  "a" => 1

julia> filter(isequal("a"), keys(d))
Set{String} with 1 element:
  "a"

julia> ds=SortedDict(d)
SortedDict{String, Int64, Base.Order.ForwardOrdering} with 2 entries:
  "a" => 1
  "b" => 2

julia> filter(isequal("a"), keys(ds))
ERROR: MethodError: no method matching filter(::Base.Fix2{typeof(isequal), String}, ::DataStructures.SDMKeyIteration{SortedDict{String, Int64, Base.Order.ForwardOrdering}})
Closest candidates are:
  filter(::Any, ::Tuple) at ~/bin/julia-1.7.1/share/julia/base/tuple.jl:363
  filter(::Any, ::Array{T, N}) where {T, N} at ~/bin/julia-1.7.1/share/julia/base/array.jl:2479
  filter(::Any, ::AbstractDict) at ~/bin/julia-1.7.1/share/julia/base/abstractdict.jl:431
  ...
Stacktrace:
 [1] top-level scope
   @ REPL[105]:1
```

**Comment by bjarthur** [comment]
perhaps because `keys` of a Dict is a Set while that of a SortedDict is an iterator:

```
julia> keys(d)
KeySet for a Dict{String, Int64} with 2 entries. Keys:
  "b"
  "a"

julia> keys(ds)
DataStructures.SDMKeyIteration{SortedDict{String, Int64, Base.Order.ForwardOrdering}}(SortedDict("a" => 1, "b" => 2))
```

**Comment by StephenVavasis** [comment]
Yes, the issue is that there is no `Base.filter` method defined for generators.  However, there is a method `Iterators.filter` that perhaps solves your problem.

**Comment by StephenVavasis** [comment]
Closed via #787
